pandas/tests/frame/indexing/test_where.py

```diff
@@ -771,3 +771,13 @@ def test_where_non_keyword_deprecation():
         result = s.where(s > 1, 10, False)
     expected = DataFrame([10, 10, 2, 3, 4])
     tm.assert_frame_equal(expected, result)
+
+
+def test_where_columns_casting():
+    # GH 42295
+
+    d1 = DataFrame({"a": [1.0, 2.0], "b": [3, np.nan]})
```

Comment:
Please call df for consistency and create expected = df.copy() before applying where

Reply:
Please create expected before calling where.